libpdbg/device.c

```diff
@@ -657,6 +657,67 @@ static void dt_link_virtual(struct pdbg_target *node, struct pdbg_target *vnode)
 	vnode->vnode = node;
 }
 
+/**
+ * @brief A function to delete the children from the list
+ * 
+ * This function is called for a valid target/node to delete the children
+ * from the list. Should be only called internally. Recursive in nature
+ * 
+ * @param target The target node for which the children will be reset
+ * 
+ * @see pdbg_reset_dt_root for more details
+ */
+static void pdbg_reset_children(struct pdbg_target* target)
+{
+	struct pdbg_target* childTraget = NULL;
+	enum pdbg_target_status status;
+
+	/* Does this target actually exist? */
+	if (target)
+	{
+		status = pdbg_target_status(target);
+		if (status != PDBG_TARGET_ENABLED)
+			return;
+	}
+	pdbg_for_each_child_target(target, childTraget)
+	{			
+		if (childTraget)
+		{
+			pdbg_reset_children(childTraget);
+		}
+		if (!list_empty(&target->children) && target->class_link.next && target->class_link.prev)
+		{
+			list_del_from(&target->children, &target->class_link);
```

Comment:
i think we need to free(&target->class_link) as it has been detached from the linked list, not childTarget as shown in line 693 to 697.

do check how many targets were present and how many were freed or else we will be leaking memory

Reply:
> i think we need to free(&target->class_link) as it has been detached from the linked list, not childTarget as shown in line 693 to 697.
> 
> do check how many targets were present and how many were freed or else we will be leaking memory

I checked and confirmed. as it is recursive in nature thus it starts freeing it up from nature. Also the class link is a list node which is a stack variable. As it points to self for its prev & next thus the condition checking `if (!list_empty(&target->children) && target->class_link.next && target->class_link.prev)` makes sure that it is free at the end of the procedure